### PR TITLE
i18n Setup : Simpler configuration + examples

### LINKED ISSUE #1590: yarn rw setup i18n fails
Running the cli command `yarn rw setup i18n` fails at the "Adding locale file for 'site' namespace" step with the error: `The "data" argument must be of type string or an instance of Buffer, TypedArray, or DataView. Received undefined`

Issue is you are missing the "contents" argument to your library packages/cli/src/commands/setup/i18n/i18n.js:  `writeFile()` method ( a wrapper to Node's `fs.writeFileSync()` which expects a contents arg). I assume your intention was to simply create an emtpy site.json file and in that case try: 

`const fd = fs.openSync(<filePath>, 'w')`

- the "w" flag makes sure the file is created if not existing, and if the file exists it overwrites it with a new file, overriding its content.
- use the "a" flag to avoid overwriting. The file is still created if not existing. 

Cheers.

### PR DESCRIPTION
**Pull request by simoncrypta**
### This is to make internalization easy with Redwood

- Replaces the i18n.js.template with a simpler and minimal configuration with examples. 
- Show a configuration for English and French translation that can easily be changed.
- Show example with a page in the comment on i18n.js file
- Error handling inspire by the tailwind setup command
- fixes #1590 

###  what's next
I think we should make a cookbook for translation at scale with RedwoodJS; I will try to work on it soon. 

We could make it work with Prerender, It's currently don't work because the useTranslate hook I guess.

We will need to make a way with Redwood Router to add the language inside the URL.

### CODE REVIEW

**Comment by dac09** [comment]
Hey @simoncrypta, thank you for this! Do you think we can get this out of draft soon? To help you with your outstanding points:
- check if if x already exists
If you use the writeFIle method, and pass the "force" flag to overwriteExisting (like you have in some places) that should cover it

-  I don't know why but new files in the templates folder cannot be found in my test
Not sure why this is happening. Are you running `build:watch` or `build` before runing test?

**Comment by simoncrypta** [comment]
Hey @dac09 , Thank you for your help! I currently search for the best way to do internalization with RedwoodJS, and I have more research to do. That's why I didn't get this out of draft yet. I plan to make internalization easy with Redwood with a good convention because they have so many ways to do it with i18n.

>   Not sure why this is happening. Are you running `build:watch` or `build` before runing test?
I running `build:watch` and everything updates except for what is on the templates folder. I would try again and write more details if isn't working again.

**Comment by thedavidprice** [comment]
@simoncrypta possibly helpful reference article from the Forums: https://community.redwoodjs.com/t/article-about-i18n-and-l10n/1477

And thanks for working on this. Huge help!

**Comment by simoncrypta** [comment]
@dac09 I think you could maybe help me to make it work with prerendering. i18next have a [doc for SSR](https://react.i18next.com/latest/ssr) , but don't understand where they but their code for _Passing initial translations / initial language down to client_ 😅

**Comment by thedavidprice** [comment]
@clairefro curious if you have availability to help review this one, specifically about high-level setup and if this includes enough foundational elements to get a project started.

**Comment by clairefro** [comment]
Cool! I will take a closer look at this tonight

> I currently search for the best way to do internalization with RedwoodJS, and I have more research to do.

In the meantime @simoncrypta , Gatsby's might be an interesting reference for how a framework handles i18n 
https://www.gatsbyjs.com/docs/how-to/adding-common-features/localization-i18n/

File: packages/cli/src/commands/setup/i18n/templates/fr.json.template [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+  "Welcome to RedwoodJS": "Bienvenu sur RedwoodJS",
+  "info": "Ceci est votre fichier de traduction",
+  "see": "https://www.i18next.com/translation-function/essentials",
+  "HomePage": {
+    "title": "Page d'accueil",
+    "info": "Touve moi dans",
```
Comment (clairefro) [inline packages/cli/src/commands/setup/i18n/templates/fr.json.template:7]:
```suggestion
    "info": "Trouve moi dans",
```
^ i don't speak French but was this a typo?

**Comment by clairefro** [comment]
Take my thought dump with a grain of salt as I'm way outta my league here (I've never peeked inside the CLI dev-side until now). I'm not familiar with Redwood's existing i18n conventions, and I didn't even know there was a `yarn rw setup i18n` command! 

i18n is a big can of worms. To touch on a few mentioned above:

**Routing**
Two common routing types are sub-path routing (`foo.com/fr/`) and domain routing (`foo.fr/`). 

Next allows both the above routing types. I highly recommend their i18n RFC thread as a reference for understanding things to be aware of, including hairy stuff like path segment localization:
https://github.com/vercel/next.js/discussions/17078

In the case of sub-path routing, oftentimes frameworks allow setting a default locale (ex: if English were the default locale: `foo.com`, not `foo.com/en`). [gatsby-plugin-i18n](https://github.com/angeloocana/gatsby-plugin-i18n) is an example of offering this 

**SEO**
I'm guessing this will rely on prerender. In addition to obvious things like localized strings for description, title etc, ideally the `html lang` attribute (ex: `<html lang="fr">`) and [hreflang attribute](https://ahrefs.com/blog/hreflang-tags/) should be set per locale to promote discoverability.

...

Lots to think about! i18n could go in so many directions - but if RW is about convention over configuration there is room to make some opinionations! Just tossing some references for now.
(another interesting read is the [Docusaurus2 i18n RFC](https://github.com/facebook/docusaurus/issues/3317) - it's a docs framework so there is a lot about markdown localization, but it also touches on valuable Routing topics)

 Love your idea to simplify the process by adding deletable sample json files and a cookbook.

File: packages/cli/src/commands/setup/i18n/templates/fr.json.template (hunk 1 above)
Comment (simoncrypta) [reply packages/cli/src/commands/setup/i18n/templates/fr.json.template:7]: Yes is a typo! Thank you, I didn't see it!

File: packages/cli/src/commands/setup/i18n/i18n.js [hunk 2]
```diff
@@ -19,56 +19,154 @@ export const builder = (yargs) => {
...
          * Add i18n import to the top of App.js
+         *
+         * Check if i18n import already exists.
+         * If it exists, throw an error.
          */
         let indexJS = fs.readFileSync(INDEX_JS_PATH)
-        indexJS = [`import './i18n'`, indexJS].join(`\n`)
-        fs.writeFileSync(INDEX_JS_PATH, indexJS)
+        if (i18nImportsExist(indexJS)) {
+          task.skip('Imports already exist in index.css')
```
Comment (thedavidprice) [inline packages/cli/src/commands/setup/i18n/i18n.js:165]:
```suggestion
          task.skip('Import already exists in App.js')
```

Simple copy edit

File: packages/cli/src/commands/setup/i18n/i18n.js [hunk 3]
```diff
@@ -19,56 +19,146 @@ export const builder = (yargs) => {
...
+         * If it exists, throw an error.
          */
-        let indexJS = fs.readFileSync(INDEX_JS_PATH)
-        indexJS = [`import './i18n'`, indexJS].join(`\n`)
-        fs.writeFileSync(INDEX_JS_PATH, indexJS)
+        let appJS = fs.readFileSync(APP_JS_PATH)
+        if (i18nImportsExist(appJS)) {
+          task.skip('Import already exists in App.js')
+        } else {
+          appJS = [`import './i18n'`, appJS].join(`\n`)
```
Comment (thedavidprice) [inline packages/cli/src/commands/setup/i18n/i18n.js:159]: I understand what's happening here, but it feels strange to me to have an import with no reference or instantiation in the code. Maybe this is commonly done and best practice as is. But I might suggest including a comment explaining what's happening with the import.

Comment (thedavidprice) [reply packages/cli/src/commands/setup/i18n/i18n.js:159]: @dac09 I was curious about whether or not this works with pretender (server-side). And it does! I tested with the example i18n code for the Home page. Just an FYI.

**Comment by simoncrypta** [comment]
Thank you @thedavidprice to fix my mistake 😅

Very interesting i18n RFC @clairefro ! Sub-path routing is definitely something essential that I didn't completely convert with `i18next-browser-languagedetector` package. Also for the `html lang` attribute, I didn't think of it and I presume is not something too hard to cover; I will work it!

**Review by thedavidprice — APPROVED** [review]
I had a couple things I wanted to test locally (e.g. if the "sync Yarn" step was necessary as it was for Tailwind; in this case it was not). And also was generally curious about how React i18n works. I found a path bug I fixed and also updated some vars to match the switch to App.js. Overall was in great shape and works really well. See my simple changes in the commits for reference.

**Comment by thedavidprice** [comment]
Thanks for the review and suggestions @clairefro Helpful indeed!

And great work here @simoncrypta This is a really helpful foundational, first step to get people up and running (and not just because it fixes the current bug).

I'll hand off to @dac09  for merge timing, but this has my approval.

Per Claire's suggestion:
> I'm guessing this will rely on prerender. In addition to obvious things like localized strings for description, title etc, ideally the html lang attribute (ex: <html lang="fr">) and hreflang attribute should be set per locale to promote discoverability.

SEO is an important addition to this initial setup and does _not_ require prerender to get started. Using something like React Helmet, we might be able to add a few pieces to this foundation that addresses both HTML lang and hreflang attributes. 

Might you be interested in exploring this as a next step here via a PR, Simon?

File: packages/cli/src/commands/setup/i18n/i18n.js (hunk 3 above)
Comment (simoncrypta) [reply packages/cli/src/commands/setup/i18n/i18n.js:159]: @thedavidprice I agree with you, but if we add a dependency to `import './index.css` and someone removes this file to have a like differents CSS files, he will not be able to make the i18n setup. What do you think?

**Comment by clairefro** [comment]
> SEO is an important addition to this initial setup and does not require prerender to get started. Using something like React Helmet, we might be able to add a few pieces to this foundation that addresses both HTML lang and hreflang attributes.

Even with React Helmet, won't prerender be necessary in order to support web crawlers for search engines and social shares?

**Comment by thedavidprice** [comment]
> Even with React Helmet, won't prerender be necessary in order to support web crawlers for search engines and social shares?

There's a lot of misconceptions here, so it's more complicated/confusing than it needs to be (which is partly intentional by Google). Crawlers render JS just fine and have for a while, so that's not really an issue. BUT site performance (alas, Lighthouse...) plays a role in the _search ranking_ performance of a website. Meaning that even if your SEO is getting indexed just fine, your site might have terrible (or amazing) end-user performance — Google will adjust where you site displays in search results accordingly. There's a lot of tricks to get sites to perform. One of them is serving static, i.e. cached, content.

Social Shares depend on OG data in the page meta, which _does_ require being rendered server-side.

Messy mess.

**tl;dr:**
Getting started with React Helmet will indeed improve SEO as is even without pre-rendered pages, which is a great start.

File: packages/cli/src/commands/setup/i18n/i18n.js (hunk 3 above)
Comment (simoncrypta) [reply packages/cli/src/commands/setup/i18n/i18n.js:159]: I find out that we can simply add it two lines up of ` const App = () => (`

Comment (thedavidprice) [reply packages/cli/src/commands/setup/i18n/i18n.js:159]: @simoncrypta sorry I missed this comment until now. 
> I find out that we can simply add it two lines up of const App = () => (

^^ This is a better approach I believe. It could also have a fallback to add at top of the file.

I'd like to see this merged for the version 0.27.0 release tomorrow. Would you be able to make this update today? 

Note: I just handled a merge conflict due to the Redwood new project template source now being TypeScript. This means `App.js` could also be `App.tsx`, so now there's a new helper to check `getPaths().web.app`

Comment (simoncrypta) [reply packages/cli/src/commands/setup/i18n/i18n.js:159]: @thedavidprice I will do my best to make this update today after my day job!

Comment (thedavidprice) [reply packages/cli/src/commands/setup/i18n/i18n.js:159]: Thanks, just keep me posted either way. I'm happy to help if/as needed.

If you do make the update, please do let me know this is ready!

Comment (simoncrypta) [reply packages/cli/src/commands/setup/i18n/i18n.js]: It was a cool challenge!

File: packages/cli/src/commands/setup/i18n/i18n.js [hunk 4]
```diff
@@ -66,6 +66,16 @@ export const handler = async ({ force }) => {
               ])
             },
           },
+          {
+            title: 'Sync yarn.lock and node_modules',
```
Comment (simoncrypta) [inline packages/cli/src/commands/setup/i18n/i18n.js:70]: @thedavidprice  I know you remove the sync Yarn because it was not necessary 
> I had a couple things I wanted to test locally (e.g. if the "sync Yarn" step was necessary as it was for Tailwind; in this case it was not).

But in my local test, I have the error `Module not found: Error: Can't resolve i18n ...` it probably happens when you use --force option 
Also, I think it's nice to have the module already install after the setup command. What do you think?

Comment (thedavidprice) [reply packages/cli/src/commands/setup/i18n/i18n.js:70]: > But in my local test, I have the error Module not found: Error: Can't resolve i18n ...

Hmm, I tested several cases locally and had no errors. In Tailwind, where this originated, after installing we used the Tailwind CLI to set up config, which needed the `yarn --check-files` in the case of --force.

I'll take a look now and see if I can confirm success.

**Comment by thedavidprice** [comment]
Great work on all this @simoncrypta 🚀

Would you be interested in opening a new Issue to discuss possible improvements regarding React Helmet + SEO?

File: packages/cli/src/commands/setup/i18n/i18n.js (hunk 4 above)
Comment (simoncrypta) [reply packages/cli/src/commands/setup/i18n/i18n.js]: I also use `yarn rwt link path/to/project` but it's probably something weird in my local setup. 😋
